### PR TITLE
Implement tests for CLI

### PR DESCRIPTION
**Pull request by fyrchik**
This is part of initial work towards covering CLI with tests #1357.

1. Use `cli.App.Writer` for the output.
2. Write some basic tests.

### CODE REVIEW

File: cli/smartcontract/smart_contract.go [hunk 1]
```diff
@@ -679,7 +679,7 @@ func getAccFromContext(ctx *cli.Context) (*wallet.Account, error) {
 		return nil, cli.NewExitError(fmt.Errorf("wallet contains no account for '%s'", address.Uint160ToString(addr)), 1)
 	}
 
-	rawPass, err := input.ReadPassword(ctx.App.Writer,
+	rawPass, err := input.ReadPassword(ctx.App.ErrWriter,
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:682]: That's just not correct, stderr is for errors.

File: cli/input/input.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"syscall"
+
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+var Input *bytes.Buffer
```
Comment (roman-khimov) [inline cli/input/input.go:15]: You don't need it.

Comment (roman-khimov) [reply cli/input/input.go:15]: Create a proper VTY and use that.

File: cli/options/options.go [hunk 3]
```diff
@@ -37,6 +37,7 @@ var RPC = []cli.Flag{
 	cli.BoolFlag{Name: "privnet, p"},
 	cli.BoolFlag{Name: "mainnet, m"},
 	cli.BoolFlag{Name: "testnet, t"},
+	cli.BoolFlag{Name: "unittest", Hidden: true},
```
Comment (roman-khimov) [inline cli/options/options.go:40]: Not sure it's a good idea, hidden parameters aren't very good in general (unless of course we're talking about super cow powers).

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: cli/executor_test.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	// Err contains command errors.
+	Err *bytes.Buffer
+}
+
+// getFreePort returns new tcp port available to listen on.
+func getFreePort(t *testing.T) uint16 {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	l, err := net.ListenTCP("tcp", addr)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply cli/executor_test.go:52]: This required a bit of refactoring. Fixed.

File: cli/smartcontract/smart_contract.go (hunk 1 above)
Comment (fyrchik) [reply cli/smartcontract/smart_contract.go:682]: Fixed.

File: cli/options/options.go (hunk 3 above)
Comment (fyrchik) [reply cli/options/options.go:40]: If we are to have `unittest` network, `Hidden` is appropriate, as it is our internal chain.
Do you want to use `privnet` instead?

File: cli/input/input.go (hunk 2 above)
Comment (fyrchik) [reply cli/input/input.go:15]: Fixed.

File: cli/executor_test.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	e.CLI.ErrWriter = e.Err
+	if needChain {
+		e.Chain, e.RPC = newTestChain(t)
+	}
+	return e
+}
+
+func (e *executor) Close() {
+	e.Chain.Close()
+	_ = e.RPC.Close()
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply cli/executor_test.go:115]: Fixed.

File: cli/executor_test.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	"github.com/urfave/cli"
+	"go.uber.org/zap/zaptest"
+)
+
+const (
+	validatorWif  = "KxyjQ8eUa4FHt3Gvioyt1Wz29cTUrE4eTqX3yFSk1YFCsPL8uNsY"
+	validatorAddr = "NVNvVRW5Q5naSx2k2iZm7xRgtRNGuZppAK"
+)
+
+var validatorHash, _ = address.StringToUint160("NVNvVRW5Q5naSx2k2iZm7xRgtRNGuZppAK")
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply cli/executor_test.go:33]: Fixed.

File: cli/executor_test.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	// NetSrv is a network server (can be empty).
+	NetSrv *network.Server
+	// Out contains command output.
+	Out *bytes.Buffer
+	// Err contains command errors.
+	Err *bytes.Buffer
+}
+
+// getFreePort returns new tcp port available to listen on.
+func getFreePort(t *testing.T) uint16 {
```
Comment (roman-khimov) [inline cli/executor_test.go:63]: Unused function?

File: cli/options/options.go (hunk 3 above)
Comment (roman-khimov) [reply cli/options/options.go:40]: Maybe we unhide it and add some comment like "don't use it"? Technically it even exists in `netmode`, so it's not that super-secret.

File: cli/wallet/validator.go [hunk 8]
```diff
@@ -101,7 +103,10 @@ func handleCandidate(ctx *cli.Context, method string) error {
 	w := io.NewBufBinWriter()
 	emit.AppCallWithOperationAndArgs(w.BinWriter, client.NeoContractHash, method, acc.PrivateKey().PublicKey().Bytes())
 	emit.Opcode(w.BinWriter, opcode.ASSERT)
-	tx, err := c.CreateTxFromScript(w.Bytes(), acc, -1, int64(gas))
+	tx, err := c.CreateTxFromScript(w.Bytes(), acc, -1, int64(gas), transaction.Signer{
+		Account: acc.Contract.ScriptHash(),
+		Scopes:  transaction.Global,
```
Comment (roman-khimov) [inline cli/wallet/validator.go:108]: `CalledByEntry`?

File: cli/wallet/validator.go [hunk 9]
```diff
@@ -156,7 +162,10 @@ func handleVote(ctx *cli.Context) error {
 	emit.AppCallWithOperationAndArgs(w.BinWriter, client.NeoContractHash, "vote", addr.BytesBE(), pubArg)
 	emit.Opcode(w.BinWriter, opcode.ASSERT)
 
-	tx, err := c.CreateTxFromScript(w.Bytes(), acc, -1, int64(gas))
+	tx, err := c.CreateTxFromScript(w.Bytes(), acc, -1, int64(gas), transaction.Signer{
+		Account: acc.Contract.ScriptHash(),
+		Scopes:  transaction.Global,
```
Comment (roman-khimov) [inline cli/wallet/validator.go:167]: `CalledByEntry`?

File: pkg/network/server.go [hunk 10]
```diff
@@ -856,6 +856,7 @@ func (s *Server) relayBlocksLoop() {
 // verifyAndPoolTX verifies the TX and adds it to the local mempool.
 func (s *Server) verifyAndPoolTX(t *transaction.Transaction) RelayReason {
 	if err := s.chain.PoolTx(t); err != nil {
+		fmt.Println(err)
```
Comment (roman-khimov) [inline pkg/network/server.go:859]: Beep

File: pkg/wallet/account.go [hunk 11]
```diff
@@ -98,6 +98,10 @@ func (a *Account) SignTx(t *transaction.Transaction) error {
 	if a.privateKey == nil {
 		return errors.New("account is not unlocked")
 	}
+	if len(a.Contract.Parameters) == 0 {
```
Comment (roman-khimov) [inline pkg/wallet/account.go:101]: If not then we don't know how to create an invocation script?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: cli/executor_test.go (hunk 7 above)
Comment (fyrchik) [reply cli/executor_test.go:63]: Fixed.

File: pkg/consensus/consensus.go [hunk 12]
```diff
@@ -189,9 +193,16 @@ func (s *service) Start() {
...
+// Shutdown implements Service interface.
+func (s *service) Shutdown() {
+	close(s.quit)
+}
+
 func (s *service) eventLoop() {
 	for {
 		select {
+		case <-s.quit:
+			return
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/consensus/consensus.go:205]: Fixed.

File: pkg/network/server.go (hunk 10 above)
Comment (fyrchik) [reply pkg/network/server.go:859]: Fixed.

File: cli/wallet/validator.go (hunk 8 above)
Comment (fyrchik) [reply cli/wallet/validator.go:108]: Fixed.

File: cli/wallet/validator.go (hunk 9 above)
Comment (fyrchik) [reply cli/wallet/validator.go:167]: Fixed.

File: cli/options/options.go (hunk 3 above)
Comment (fyrchik) [reply cli/options/options.go:40]: `don't use it` is exactly the case for hidden parameters. IMO `--unittest` will just overload help message.

File: pkg/wallet/account.go (hunk 11 above)
Comment (fyrchik) [reply pkg/wallet/account.go:101]: Yes, we don't yet have a proper CLI and haven't discussed what cases to support.

File: pkg/rpc/server/server.go [hunk 13]
```diff
@@ -174,11 +174,19 @@ func (s *Server) Start(errChan chan error) {
...
 	}
-	err := s.ListenAndServe()
-	if err != http.ErrServerClosed {
-		s.log.Error("failed to start RPC server", zap.Error(err))
+	ln, err := net.Listen("tcp", s.Addr)
+	if err != nil {
 		errChan <- err
+		return
 	}
+	s.Addr = ln.Addr().String() // set Addr to the actual address
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/rpc/server/server.go:182]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Comment by roman-khimov** [comment]
Please update it to fix merge conflict.

**Pull request merged**